File: dispatch-connect/samples/Java/DispatchOutbound.java
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.databind.ObjectMapper;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.List;

public class DispatchOutbound {

    public static void main(String[] args) {
        try {
            String PublicKey = "your_public_id";
            String SecretKey = "your_secret_key";
            String maxMessages = "{\"maxNumberOfMessages\": 10}";

            int numRecs = 0;
            // Calculate Dispatch Signature
            byte[] bMaxMessages = maxMessages.getBytes(java.nio.charset.StandardCharsets.UTF_8);
            byte[] bSecretKey = ConvertHexStringToByteArray(SecretKey);

            HttpURLConnection AgentOutEndPoint = (HttpURLConnection) new URL("https://connect-sbx.dispatch.me/agent/out").openConnection();
            AgentOutEndPoint.setRequestMethod("POST");
            AgentOutEndPoint.setRequestProperty("Content-Type", "application/json");
            AgentOutEndPoint.setRequestProperty("X-Dispatch-Key", PublicKey);
            AgentOutEndPoint.setDoOutput(true);
            DataOutputStream AgentOutStream;
            do {
                String OutSignature = GetSignatureHash(bSecretKey, bMaxMessages);
                AgentOutEndPoint.setRequestProperty("X-Dispatch-Signature", OutSignature);
                AgentOutStream = new DataOutputStream(AgentOutEndPoint.getOutputStream());
                AgentOutStream.write(bMaxMessages);

                BufferedReader reader = new BufferedReader(new InputStreamReader(AgentOutEndPoint.getInputStream()));
                StringBuilder stringBuilder = new StringBuilder();
                String line = null;
                while ((line = reader.readLine()) != null)
                {
                    stringBuilder.append(line + "\n");
                }
                String AgentOutString = stringBuilder.toString();
                ObjectMapper objectMapper = new ObjectMapper();

                AgentOutJson[] AgentOut = objectMapper.readValue(AgentOutString, AgentOutJson[].class);
                MessageJson m;
                numRecs = AgentOut.length;
                for (AgentOutJson rec : AgentOut) {
                    m = rec.Message;
                    String ret = "Success";
                    String err = "";
                    try
                    {
                        JobJson payload = m.Request.Payload.Actions.get(0).Put.job;
                        ////////////////////////////////////////////////////////////
                        //  DO SOMETHING HERE WITH THE PAYLOAD - update your system!
                        ////////////////////////////////////////////////////////////
                    }
                    catch (Exception e)
                    {
                        ret = "error";
                        err = e.getMessage();
                    }
                    finally
                    {
                        String receipt = String.format("%s\"Receipt\":\"%s\",\"ProcedureID\":\"%s\",\"Result\":\"%s\",\"Error\":\"%s\"%s", "{", m.Receipt, m.Request.ProcedureID, ret, err, "}");
                        byte[] bReceipt = receipt.getBytes(java.nio.charset.StandardCharsets.UTF_8);
                        String AckSignature = GetSignatureHash(bSecretKey, bReceipt);

                        HttpURLConnection AgentAckEndPoint = (HttpURLConnection) new URL("https://connect-sbx.dispatch.me/agent/ack").openConnection();
                        AgentAckEndPoint.setRequestMethod("POST");
                        AgentAckEndPoint.setRequestProperty("Content-Type", "application/json");
                        AgentAckEndPoint.setRequestProperty("X-Dispatch-Key", PublicKey);
                        AgentAckEndPoint.setRequestProperty("X-Dispatch-Signature", AckSignature);
                        AgentAckEndPoint.setDoOutput(true);

                        DataOutputStream AgentAckStream = new DataOutputStream(AgentAckEndPoint.getOutputStream());
                        AgentAckStream.write(bReceipt);
                        // System.out.println(AgentAckEndPoint.getResponseMessage());
                    }
                }
            } while (numRecs >= 10);
            AgentOutStream.flush();
            AgentOutStream.close();
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    }

    private static byte[] ConvertHexStringToByteArray(String hexString)
    {
        int len = hexString.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2){
            data[i / 2] = (byte) ((Character.digit(hexString.charAt(i), 16) << 4)
                    + Character.digit(hexString.charAt(i+1), 16));
        }
        return data;
    }

    private static String GetSignatureHash(byte[] key, byte[] message) throws Exception
    {
        byte[] returnVal = null;
        String algorithm="HmacSHA256";
        SecretKeySpec signingKey = new SecretKeySpec(key, algorithm);
        Mac mac = Mac.getInstance(algorithm);
        mac.init(signingKey);
        returnVal = mac.doFinal(message);
        StringBuilder hashString = new StringBuilder();
        for (byte x : returnVal)
        {
            hashString.append(String.format("%02x", x));
        }
        return hashString.toString();
    }
}

@JsonIgnoreProperties(ignoreUnknown = true)
class AgentOutJson
{
    public MessageJson Message;
}

@JsonIgnoreProperties(ignoreUnknown = true)
class MessageJson
{
    public RequestJson Request;
    public String ID;
    public String Receipt;
}

@JsonIgnoreProperties(ignoreUnknown = true)
class RequestJson
{
    public RequestJson() {}
    public String ProcedureID;
    public String Type;
    public PayloadJson Payload;
}

@JsonIgnoreProperties(ignoreUnknown = true)
class PayloadJson
{
    public List<ActionsJson> Actions;
}

@JsonIgnoreProperties(ignoreUnknown = true)
class ActionsJson
{
    public String ID;
    public PutJson Put;
}

@JsonIgnoreProperties(ignoreUnknown = true)
class PutJson
{
    public JobJson job;
}

@JsonIgnoreProperties(ignoreUnknown = true)
class JobJson
{
    public int id;
    public String title;
    public String created_at;
    public String updated_at;
    public AddressJson address;
    public CustomerJson customer;
    public List<EquipmentJson> equipment_descriptions;
    public List<String> external_ids;
    public String service_type;
    public float service_fee;
    public String symptom;
    public String timezone;
    public CustomFieldsJson custom_fields;
}

@JsonIgnoreProperties(ignoreUnknown = true)
class AddressJson
{
    public String street_1;
    public String street_2;
    public String postal_code;
    public String city;
    public String state;
    public String country;
    public double latitude;
    public double longitude;
}

@JsonIgnoreProperties(ignoreUnknown = true)
class CustomerJson
{
    public String first_name;
    public String last_name;
    public String email;
    public String city;
    public String phone_number;
}

@JsonIgnoreProperties(ignoreUnknown = true)
class EquipmentJson
{
    public String equipment_type;
    public String installation_date;
    public String manufacturer;
    public String model_number;
    public String serial_number;
}

@JsonIgnoreProperties(ignoreUnknown = true)
class CustomFieldsJson
{
    public String custom_field_1;
    public String custom_field_2;
}

File: dispatch-connect/samples/CSharp/DispatchOutbound.cs
using System;
using System.Globalization;
using System.IO;
using System.IO.Compression;
using System.Net;
using System.Security.Cryptography;
using System.Text;
using Newtonsoft.Json;

namespace DispatchConnect
{
    class Program
    {

        public class AgentOutJson
        {
            public MessageJson Message { get; set; }
        }

        public class MessageJson
        {
            public RequestJson Request { get; set; }
            public String ID { get; set; }
            public String Receipt { get; set; }
        }

        public class RequestJson
        {
            public String ProcedureID { get; set; }
            public String Type { get; set; }
            public PayloadJson Payload { get; set; }
        }

        public class PayloadJson
        {
            public ActionsJson[] Actions { get; set; }
        }

        public class ActionsJson
        {
            public String ID { get; set; }
            public PutJson Put { get; set; }
        }

        public class PutJson
        {
            public JobJson Job { get; set; }
        }

        public class JobJson
        {
            public int id { get; set; }
            public String title { get; set; }
            public String created_at { get; set; }
            public String updated_at { get; set; }
            public AddressJson address { get; set; }
            public CustomerJson customer { get; set; }
            public EquipmentJson[] equipment_descriptions { get; set; }
            public String[] external_ids { get; set; }
            public String service_type { get; set; }
            public float service_fee { get; set; }
            public float service_fee_pre_collected { get; set; }
            public String symptom { get; set; }
            public String timezone { get; set; }
            public CustomFieldsJson custom_fields { get; set; }

        }

        public class AddressJson
        {
            public String street_1 { get; set; }
            public String street_2 { get; set; }
            public String postal_code { get; set; }
            public String city { get; set; }
            public String state { get; set; }
            public String country { get; set; }
            public double latitude { get; set; }
            public double longitude { get; set; }
        }

        public class CustomerJson
        {
            public String first_name { get; set; }
            public String last_name { get; set; }
            public String email { get; set; }
            public String city { get; set; }
            public String phone_number { get; set; }
        }

        public class EquipmentJson
        {
            public String equipment_type { get; set; }
            public String installation_date { get; set; }
            public String manufacturer { get; set; }
            public String model_number { get; set; }
            public String serial_number { get; set; }
        }

        public class CustomFieldsJson
        {
            public String custom_field_1 { get; set; }
            public String custom_field_2 { get; set; }
        }

        const string PublicKey = "your_public_key";
        const string SecretKey = "your_secret";
        const string maxMessages = "{\"maxNumberOfMessages\": 10}";
        static void Main(string[] args)
        {
            int numRecs = 0;
            byte[] bMaxMessages = Encoding.UTF8.GetBytes(maxMessages);
            byte[] mysecretyKey = ConvertHexStringToByteArray(SecretKey);
            HttpWebRequest AgentOutEndPoint = (HttpWebRequest)HttpWebRequest.Create("https://connect-sbx.dispatch.me/agent/out");   // the production API is https://connect.dispatch.me
            AgentOutEndPoint.Method = "POST";
            AgentOutEndPoint.Headers.Add("X-Dispatch-Key", PublicKey);
            AgentOutEndPoint.ContentType = "application/json";
            do
            {
                string OutSignature = GetSignatureHash(mysecretyKey, bMaxMessages);
                AgentOutEndPoint.Headers.Add("X-Dispatch-Signature", OutSignature);
                Stream outStream = AgentOutEndPoint.GetRequestStream();
                outStream.Write(bMaxMessages, 0, bMaxMessages.Length);
                HttpWebResponse ResponseAgentOut = (HttpWebResponse)AgentOutEndPoint.GetResponse();
                Stream receiveOutStream = ResponseAgentOut.GetResponseStream();
                StreamReader readOutStream = new StreamReader(receiveOutStream, Encoding.UTF8);

                String json = readOutStream.ReadToEnd();
                AgentOutJson[] agentOut = JsonConvert.DeserializeObject<AgentOutJson[]>(json);
                MessageJson m;
                numRecs = agentOut.Length;
                foreach (AgentOutJson rec in agentOut)
                {
                    m = rec.Message;
                    String ret = "Success";
                    String err = "";
                    try
                    {
                        JobJson payload = m.Request.Payload.Actions[0].Put.Job;
                        ////////////////////////////////////////////////////////////
                        //  DO SOMETHING HERE WITH THE PAYLOAD - update your system!
                        ////////////////////////////////////////////////////////////
                    }
                    catch (Exception e)
                    {
                        ret = "error";
                        err = e.Message;
                    }
                    finally
                    {
                        String receipt = string.Format("{0}\"Receipt\":\"{1}\",\"ProcedureID\":\"{2}\",\"Result\":\"{3}\",\"Error\":\"{4}\"{5}", "{", m.Receipt, m.Request.ProcedureID, ret, err, "}");
                        byte[] bReceipt = Encoding.UTF8.GetBytes(receipt);
                        String AckSignature = GetSignatureHash(mysecretyKey, bReceipt);
                        HttpWebRequest AgentAckEndPoint = (HttpWebRequest)HttpWebRequest.Create("https://connect-sbx.dispatch.me/agent/ack");
                        AgentAckEndPoint.Method = "POST";
                        AgentAckEndPoint.Headers.Add("X-Dispatch-Signature", AckSignature);
                        AgentAckEndPoint.Headers.Add("X-Dispatch-Key", PublicKey);
                        AgentAckEndPoint.ContentType = "application/json";
                        Stream ackStream = AgentAckEndPoint.GetRequestStream();
                        ackStream.Write(bReceipt, 0, bReceipt.Length);
                        HttpWebResponse ResponseAgentAck = (HttpWebResponse)AgentAckEndPoint.GetResponse();
                        Stream receiveAckStream = ResponseAgentAck.GetResponseStream();
                        StreamReader readAckstream = new StreamReader(receiveAckStream, Encoding.UTF8);
                    }

                }
            } while (numRecs >= 10);
        }

        public static byte[] Compress(byte[] raw)
        {
            using (MemoryStream memory = new MemoryStream())
            {
                using (GZipStream gzip = new GZipStream(memory, CompressionMode.Compress, true))
                {
                    gzip.Write(raw, 0, raw.Length);
                }
                return memory.ToArray();
            }
        }
        public static byte[] ConvertHexStringToByteArray(string hexString)
        {
            byte[] HexAsBytes = new byte[hexString.Length / 2];
            for (int index = 0; index < HexAsBytes.Length; index++)
            {
                string byteValue = hexString.Substring(index * 2, 2);
                HexAsBytes[index] = byte.Parse(byteValue, NumberStyles.HexNumber, CultureInfo.InvariantCulture);
            }
            return HexAsBytes;
        }

        public static String GetSignatureHash(byte[] key, byte[] message)
        {
            HMACSHA256 hmac = new HMACSHA256(key);
            byte[] hash = hmac.ComputeHash(message);
            StringBuilder hashString = new StringBuilder();
            foreach (byte x in hash)
            {
                hashString.AppendFormat("{0:x2}", x);
            }
            return hashString.ToString();
        }
    }
}
